src/osd/osd_types.h

```diff
@@ -2744,80 +2744,360 @@ inline ostream& operator<<(ostream& out, const pg_log_t& log)
+  void get_changed(F &&f) {
+    for (auto const &i: _changed) {
+      f(i);
     }
-  }; 
-  WRITE_CLASS_ENCODER(item)
+  }
+  void flush() {
+    _changed.clear();
+  }
+  bool clean() const {
+    return _changed.empty();
+  }
+};
 
-private:
+template <bool TrackChanges>
+class pg_missing_set : public pg_missing_const_i {
+  using item = pg_missing_item;
   map<hobject_t, item, hobject_t::ComparatorWithDefault> missing;  // oid -> (need v, have v)
   map<version_t, hobject_t> rmissing;  // v -> oid
+  ChangeTracker<TrackChanges> tracker;
 
 public:
-  const map<hobject_t, item, hobject_t::ComparatorWithDefault> &get_items() const {
+  pg_missing_set() = default;
+
+  template <typename missing_type>
+  pg_missing_set(const missing_type &m) :
+    missing(m.get_items()), rmissing(m.get_rmissing()) {
+    static_assert(
+      !TrackChanges,
+      "Cannot copy construct into tracker pg_missing object");
+  }
+
+  const map<hobject_t, item, hobject_t::ComparatorWithDefault> &get_items() const override {
     return missing;
   }
-  const map<version_t, hobject_t> &get_rmissing() const {
+  const map<version_t, hobject_t> &get_rmissing() const override {
     return rmissing;
   }
-  unsigned int num_missing() const;
-  bool have_missing() const;
-  bool is_missing(const hobject_t& oid) const;
-  bool is_missing(const hobject_t& oid, eversion_t v) const;
-  eversion_t have_old(const hobject_t& oid) const;
-
-  void swap(pg_missing_t& o);
-  void add_next_event(const pg_log_entry_t& e);
-  void revise_need(hobject_t oid, eversion_t need);
-  void revise_have(hobject_t oid, eversion_t have);
-  void add(const hobject_t& oid, eversion_t need, eversion_t have);
-  void rm(const hobject_t& oid, eversion_t v);
-  void rm(std::map<hobject_t, pg_missing_t::item, hobject_t::ComparatorWithDefault>::const_iterator m);
-  void got(const hobject_t& oid, eversion_t v);
-  void got(std::map<hobject_t, pg_missing_t::item, hobject_t::ComparatorWithDefault>::const_iterator m);
-  void split_into(pg_t child_pgid, unsigned split_bits, pg_missing_t *omissing);
+  unsigned int num_missing() const override {
+    return missing.size();
+  }
+  bool have_missing() const override {
+    return !missing.empty();
+  }
+  bool is_missing(const hobject_t& oid) const override {
+    return (missing.find(oid) != missing.end());
+  }
+  bool is_missing(const hobject_t& oid, eversion_t v) const override {
+    map<hobject_t, item, hobject_t::ComparatorWithDefault>::const_iterator m =
+      missing.find(oid);
+    if (m == missing.end())
+      return false;
+    const item &item(m->second);
+    if (item.need > v)
+      return false;
+    return true;
+  }
+  eversion_t have_old(const hobject_t& oid) const override {
+    map<hobject_t, item, hobject_t::ComparatorWithDefault>::const_iterator m =
+      missing.find(oid);
+    if (m == missing.end())
+      return eversion_t();
+    const item &item(m->second);
+    return item.have;
+  }
+
+  void swap(pg_missing_set& o) {
```

Comment:
They are using it to place the incoming missing set into the peer_missing map.  *this is being modified.  Although, it should really be called claim(), and it should be disabled for the variant with the tracker.